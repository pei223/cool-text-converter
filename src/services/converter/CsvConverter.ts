import { ConvertibleMarkdownFormats } from "../consts/format";
import { MarkdownUnrecognizedError } from "../errors/types";
import { writeMarkdownTable } from "./tableWriter";
import { MarkdownConverter } from "./types";

export class CsvConverter implements MarkdownConverter {
  recognize(v: string): void {
    // 1行目にカンマがあるかで判断.
    // かなり誤検知がありそうなので優先度は低めにしておく.
    const firstLine = v.split("\n")[0];
    if (!firstLine.includes(",")) {
      throw new MarkdownUnrecognizedError(v, "CSV", `Not contained comma`);
    }
  }
  getDetailedConverter(): MarkdownConverter | null {
    return null;
  }
  convertToMarkdown(v: string): string {
    const lines = v.split("\n");
    return writeMarkdownTable(
      lines[0].split(","),
      lines.length === 1 ? [] : lines.map((v) => v.split(",")).slice(1)
    );
  }
  formatName(): ConvertibleMarkdownFormats {
    return "CSV";
  }
  formatLabel(): string {
    return "CSV";
  }
}
