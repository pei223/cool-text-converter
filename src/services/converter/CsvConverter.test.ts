import { describe, it, expect } from "vitest";
import { CsvConverter } from "./CsvConverter";

describe("CsvConverter", () => {
  const converter = new CsvConverter();

  describe("recognize", () => {
    it.each(["test,test", "tett,test,test", "head,head,head\ntett,test,test"])(
      "Success(%s)",
      (v: string) => {
        expect(() => converter.recognize(v)).not.toThrow();
      }
    );
    it.each(["http//hoge", "1234", "invalid"])("Not apply(%s)", (v: string) => {
      expect(() => converter.recognize(v)).toThrow();
    });
  });
  // 中身いちいち正解かくのしんどいのでスナップショットテストにする
  it.each([
    ["test,test"],
    ["tett,test,test"],
    ["head,head,head\ntett,test,test"],
  ])("converteToMarkdown(%s) - (%s)", (v: string) => {
    expect(converter.convertToMarkdown(v)).toMatchSnapshot();
  });
});
