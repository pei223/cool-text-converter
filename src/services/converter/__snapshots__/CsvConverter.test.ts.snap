// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CsvConverter > converteToMarkdown(head,head,head
tett,test,test) - (%s) 1`] = `
"| head | head | head |
| ---- | ---- | ---- |
| tett | test | test |"
`;

exports[`CsvConverter > converteToMarkdown(test,test) - (%s) 1`] = `
"| test | test |
| ---- | ---- |
"
`;

exports[`CsvConverter > converteToMarkdown(tett,test,test) - (%s) 1`] = `
"| tett | test | test |
| ---- | ---- | ---- |
"
`;
